arangod/RestServer/DatabaseFeature.cpp

```diff
@@ -291,6 +294,165 @@ void DatabaseManagerThread::run() {
   }
 }
 
+struct HeartbeatTimescale {
+  static arangodb::metrics::LogScale<double> scale() {
+    return {10.f, 0.f, 1000000.f, 6};
+  }
+};
+
+DECLARE_HISTOGRAM(arangodb_ioheartbeat_duration, HeartbeatTimescale,
+                  "Time to execute the io heartbeat once [ms]");
+DECLARE_COUNTER(arangodb_ioheartbeat_failures_write_total,
+                "Total number of failures in IO heartbeat when writing");
+DECLARE_COUNTER(arangodb_ioheartbeat_failures_read_total,
+                "Total number of failures in IO heartbeat when reading");
+DECLARE_COUNTER(arangodb_ioheartbeat_failures_remove_total,
+                "Total number of failures in IO heartbeat when removing");
+DECLARE_COUNTER(arangodb_ioheartbeat_delays_write_total,
+                "Total number of delays in IO heartbeat when writing");
+DECLARE_COUNTER(arangodb_ioheartbeat_delays_read_total,
+                "Total number of delays in IO heartbeat when reading");
+DECLARE_COUNTER(arangodb_ioheartbeat_delays_remove_total,
+                "Total number of delays in IO heartbeat when removing");
+
+/// IO check thread main loop
+/// The purpose of this thread is to try to perform a simple IO write
+/// operation on the database volume regularly. We need visibility in
+/// production if IO is slow or not possible at all.
+IOHeartbeatThread::IOHeartbeatThread(Server& server,
+                                     metrics::MetricsFeature& metricsFeature)
+    : ServerThread<ArangodServer>(server, "IOHeartbeat"),
+      _exeTimeHistogram(metricsFeature.add(arangodb_ioheartbeat_duration{})),
+      _writeFailures(
+          metricsFeature.add(arangodb_ioheartbeat_failures_write_total{})),
+      _readFailures(
+          metricsFeature.add(arangodb_ioheartbeat_failures_read_total{})),
+      _removeFailures(
+          metricsFeature.add(arangodb_ioheartbeat_failures_remove_total{})),
+      _writeDelays(
+          metricsFeature.add(arangodb_ioheartbeat_delays_write_total{})),
+      _readDelays(metricsFeature.add(arangodb_ioheartbeat_delays_read_total{})),
+      _removeDelays(
+          metricsFeature.add(arangodb_ioheartbeat_delays_remove_total{})) {}
+
+IOHeartbeatThread::~IOHeartbeatThread() { shutdown(); }
+
+void IOHeartbeatThread::run() {
+  auto& databasePathFeature = server().getFeature<DatabasePathFeature>();
+  std::string testFilePath = FileUtils::buildFilename(
+      databasePathFeature.directory().c_str(), "TestBalloonFileIOHeartbeat");
+  std::string testFileContent = "This is just an I/O test.\n";
+
+  LOG_TOPIC("66665", INFO, Logger::ENGINES)
+      << "IOHeartbeatThread: running with checkInterval= " << checkInterval()
+      << " ...";
+
+  unsigned int counter = checkInterval();
+
+  while (true) {
+    try {  // protect thread against any exceptions
+      if (isStopping()) {
+        // done
+        break;
+      }
+
+      if (--counter == 0) {
```

Comment:
Instead of using a `counter` variable here and repeatedly waking up the thread without performing actual work, we should use a condition variable. The thread could sleep on the condition variable for 15s then.
On server shutdown, we can signal the condition variable so the thread can wake up timely. All we will need to do for this it to implement a `beginShutdown()` method for the thread that signals, and that calls the parent Thread `beginShutdown()` method. I think this is both cleaner and more efficient than the counter-based approach.

Reply:
Done.